### PR TITLE
FileDialog : Adding mounted volumes on Linux

### PR DESCRIPTION
**Pull request by MrTopom**
First version to add the mounted volumes in the FileDialog on Linux.

I based the filtering of the file system types from an online documentation. Any feedback is welcome for a better approach to define usable file systems, if there is any.

### CODE REVIEW

File: src/gui/modals/FileDialog.cpp
Comment (michaelgregorius) [inline src/gui/modals/FileDialog.cpp:95]: I propose to also add the following types:
* "ntfs3": For users who have mounted NTFS partitions under Linux.
* "fuse.sshfs": Access to remote file systems via SSH.

I have tested both types on my Linux machine.

The "fuse.sshfs" mount is a remote access to my smartphone via KDE Connect so adding this mount type would add quite some flexibility.

There's also another NTFS driver that's provided by [ntfs-3g](https://wiki.ubuntuusers.de/Archiv/FUSE/ntfs-3g/) but I don't know the exact name of the mount type right now. I thinks it's also something with fuse.

Comment (michaelgregorius) [reply src/gui/modals/FileDialog.cpp:95]: I have just checked by mounting a USB stick that's formatted to NTFS using the [ntfs-3g](https://wiki.ubuntuusers.de/Archiv/FUSE/ntfs-3g/) driver. The mount type in this case is "fuseblk". So if "ntfs3" and "fuseblk" were added then most NTFS partitions would be shown.

File: src/gui/modals/FileDialog.cpp
Comment (michaelgregorius) [inline src/gui/modals/FileDialog.cpp:105]: This line should be changed to:
```
urls << QUrl::fromLocalFile(storage.rootPath());
```
Otherwise it won't work with partitions that have names/labels. Here's a snippet of some debug output for my filesystem:
```
Display name:  "/data/linux" , Name:  "" , Root Path:  "/data/linux"
Display name:  "Linux-Homes" , Name:  "Linux-Homes" , Root Path:  "/home"
Display name:  "usb-2.0-data" , Name:  "usb-2.0-data" , Root Path:  "/mnt"
```
My `home` folder is mounted on a separate `ext4` partition that is named "Linux-Homes". If `displayName` is added to the lists of URLs then "Linux-Homes" will be shown in the sidebar. This looks nice but unfortunately it does not work when I click on it. Changing the line as described above leads to `/home`  being displayed in the sidebar. This does not look as nice but it's working.

The partition `/data/linux` does not have a name. Hence the display name and the technical root path are the same and it also works with the display name in the URL list.

Unfortunately it does not seem to be possible to differentiate in the sidebar between the name that should be displayed and the technical path where the data resides.

File: src/gui/modals/FileDialog.cpp
Comment (michaelgregorius) [inline src/gui/modals/FileDialog.cpp:103]: The two if statements can be collapsed into one statement:
```
if (storage.isValid() && storage.isReady() && usableFileSystems.contains(QString(storage.fileSystemType()), Qt::CaseInsensitive))
```
Or alternatively:
```
bool const typeInSupportedList = usableFileSystems.contains(QString(storage.fileSystemType()), Qt::CaseInsensitive);
if (storage.isValid() && storage.isReady() && typeInSupportedList)
```

**Review by michaelgregorius — APPROVED** [review]
Looks good to me.

File: src/gui/modals/FileDialog.cpp
Comment (messmerd) [inline src/gui/modals/FileDialog.cpp:54]:
```suggestion
#else
```

**Review by messmerd — COMMENTED** [review]
I'll test it sometime tomorrow

File: src/gui/modals/FileDialog.cpp
Comment (DomClark) [inline src/gui/modals/FileDialog.cpp:95]: Maybe I'm just a Windows brainlet, but why would you have a filesystem mounted that wasn't useable? I'm averse to whitelisting (or blacklisting) external things in the source code, as it's a burden we have to keep up-to-date, and our releases aren't exactly frequent. Is there actually any risk in displaying file systems whose types aren't in this list?

File: src/gui/modals/FileDialog.cpp
Comment (DomClark) [inline src/gui/modals/FileDialog.cpp:60]: Why not construct `desktopDir` directly with the path? Also, you can scope the variable to the if-statement in C++17:
```suggestion
	if (const auto desktopDir = QDir{QStandardPaths::writableLocation(QStandardPaths::DesktopLocation)};
		desktopDir.exists())
```
Similarly for the other directories here.

File: src/gui/modals/FileDialog.cpp
Comment (MrTopom) [reply src/gui/modals/FileDialog.cpp:95]: I did it only to prevent to show volumes with system filesystems like 'tmpfs'.

File: src/gui/modals/FileDialog.cpp
Comment (MrTopom) [reply src/gui/modals/FileDialog.cpp:60]: I am not aware of this coding feature and find it less readable. But if this is best practice I could change.

Comment (DomClark) [reply src/gui/modals/FileDialog.cpp:57]: Feel free to leave it if you'd rather not use it. I personally like it as it reduces the scope of the variable, which I find easier to read as I know where the variable is relevant with more precision. But the code needs to be readable by everyone, so if you're more comfortable not doing this, leave it. The scope is not particularly important in this case - it's not a mutex lock or anything like that, which is where this feature really comes in handy. I do prefer passing the path to `QDir`'s constructor though, if you're happy just to do that bit.

File: src/gui/modals/FileDialog.cpp
Comment (DomClark) [reply src/gui/modals/FileDialog.cpp:92]: Makes sense. Is tmpfs only used for `/tmp` and its ilk, or could it reasonably be seen in normal usage too? Would it be better to blacklist tmpfs rather than whitelisting the others? Should particular system directories, rather than filesystem types, be omitted instead? I don't know what's best here, so I'll leave it up to you Linux people.

**Comment by messmerd** [comment]
It works for me on Linux Mint 21.1.

File: src/gui/modals/FileDialog.cpp
Comment (michaelgregorius) [reply src/gui/modals/FileDialog.cpp:92]: A modern Linux has many technical file systems/types that are mounted while running. Here's a list of the types that are currently mounted on my system:
```
$ findmnt -lo fstype | sort | uniq
autofs
bpf
cgroup2
configfs
debugfs
devpts
devtmpfs
efivarfs
ext4
fusectl
hugetlbfs
mqueue
ntfs3
proc
pstore
securityfs
sysfs
tmpfs
tracefs
vfat
```

Comment (DomClark) [reply src/gui/modals/FileDialog.cpp:92]: I see. I'll resolve this conversation then, since it doesn't look like there's a better approach for now.

**Review by Veratil — APPROVED** [review]

**Comment by messmerd** [comment]
Coming back to this, it looks like on the master branch, FileDialog already showed my mounted volumes on Linux. I'm not sure why.

**Comment by zonkmachine** [comment]
> on the master branch, FileDialog already showed my mounted volumes on Linux. I'm not sure why.

I was about to merge this one. Is there any difference in behavior at all between your master and this PR?

**Comment by Veratil** [comment]
Could it be a Qt change?

**Comment by messmerd** [comment]
> Is there any difference in behavior at all between your master and this PR?

I didn't notice any difference.

**Comment by zonkmachine** [comment]
It would be nice to know what's going on here but is this reason to wait with merging this PR now?

**Comment by Veratil** [comment]
I would like @MrTopom to test master without this PR before we merge. I would like to understand if this is something that Qt changed or if there's another reason. If this is actually not needed then we have no reason to merge it in.

@messmerd can you share which version of Qt is linked for your build so we can compare? @MrTopom please share details as well.

**Comment by messmerd** [comment]
@Veratil I'm using Qt 5.15.2.

**Comment by michaelgregorius** [comment]
I can imagine that [setSidebarUrls](https://doc.qt.io/qt-6/qfiledialog.html#setSidebarUrls) caches/stores the URLs that are set on a per application/executable basis. In that case it would act more like a favorites feature. If that's the case it does not seem to be documented thought.

A quick test would be to add another line to LMMS' file dialog class that adds the URL to a very specific directory on the developer machine so that another entry appears in the side bar. Once it's there, delete that code and start the application again. If it is still there then the behavior is specific to [setSidebarUrls](https://doc.qt.io/qt-6/qfiledialog.html#setSidebarUrls).

**Comment by michaelgregorius** [comment]
Ok, I have just done what I have proposed and the directory is still there.

To reproduce:
1. Add the folder `~/lmms-sidebar-test`.
2. Optional: Copy a LMMS project file to `~/lmms-sidebar-test`.
3. Add the following code before the call to `setSidebarUrls(urls);`:
```
urls << QUrl::fromLocalFile(QDir(QDir::homePath() + "/lmms-sidebar-test").absolutePath());
```
4. Compile and start LMMS.
5. Open the file dialog. The entry should be there.
6. Close LMMS.
7. Comment out the line that was added in step 3.
8. Compile and start LMMS again.
9. Open the file dialog. The entry is still there. You can remove it with a right click.

**Comment by messmerd** [comment]
@michaelgregorius 
Thanks for investigating this!

I removed the mounted volumes from my sidebar, then opened a master branch build of LMMS. They were NOT added back in.
Then I started a build of this feature branch and the mounted volumes appeared in the sidebar once again.

That clears up my confusion, and confirms that the changes this PR makes are indeed necessary.

**Review by messmerd — APPROVED** [review]

**Pull request merged**
